### PR TITLE
Use strict behavior for aweek objects

### LINKED ISSUE #8: Add data.frame.aweek method
Currently, it's only possible to create aweek data frames by inserting the column. Creating the `data.frame.aweek()` method would allow users to create data frames on the fly with aweek columns:

``` r
library('aweek')
d <- as.Date("2011-02-08") + 1:10
w <- date2week(d)
data.frame(w, d)
#> Error in as.data.frame.default(x[[i]], optional = TRUE, stringsAsFactors = stringsAsFactors): cannot coerce class '"aweek"' to a data.frame
df <- data.frame(d)
df$w <- w
df
#>             d          w
#> 1  2011-02-09 2011-W06-3
#> 2  2011-02-10 2011-W06-4
#> 3  2011-02-11 2011-W06-5
#> 4  2011-02-12 2011-W06-6
#> 5  2011-02-13 2011-W06-7
#> 6  2011-02-14 2011-W07-1
#> 7  2011-02-15 2011-W07-2
#> 8  2011-02-16 2011-W07-3
#> 9  2011-02-17 2011-W07-4
#> 10 2011-02-18 2011-W07-5
# Tibbles work
tibble::tibble(w = w, d = d)
#> # A tibble: 10 x 2
#>    w           d         
#>    <S3: aweek> <date>    
#>  1 2011-W06-3  2011-02-09
#>  2 2011-W06-4  2011-02-10
#>  3 2011-W06-5  2011-02-11
#>  4 2011-W06-6  2011-02-12
#>  5 2011-W06-7  2011-02-13
#>  6 2011-W07-1  2011-02-14
#>  7 2011-W07-2  2011-02-15
#>  8 2011-W07-3  2011-02-16
#>  9 2011-W07-4  2011-02-17
#> 10 2011-W07-5  2011-02-18
```

<sup>Created on 2019-03-11 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

### LINKED ISSUE #20: Remove ambiguous behavior (breaking changes)
Things that should be changed:

1. behavior of factors
2. combining aweek objects

After thinking about it (with a bit of help from Rich FitzJohn), I think adding the ability to combine aweek objects with different start days was a mistake. The behavior of simply taking the first element and converting it was a mistake because it will not be obvious to the user what is happening. Combining aweek objects should do the following:

1. only combine aweek objects with the same week_start attribute
   i. if they don't have the same week_start attribute, an error is thrown, advising the use of `change_week_start()`
2. if combining a factor, it is converted to a character, no matter what. The user can use `factor_aweek()` (or something like that) to change it.
3. truncated aweek objects are no longer truncated.

Other than that, character and date objects are trivial

### PR DESCRIPTION
**Pull request by zkamvar**
This PR will fix #20 and fix #8 

The aweek package is going through a bit of a change to make things more
consistent. The list of changes can be found in the NEWS.md. Please test out
this code and let me know if you encounter any problems. I plan to release
this the second week of June.

The breaking changes in this release are that it’s no longer possible to
combine aweek objects that have different week\_start attributes and factors
no longer coerce factors (but will be converted to characters upon
combination.

In the previous release, I enabled global week\_start parameters that allowed
the user to set the week\_start variable globally.

Please test this out with:

```r
# install.packages("remotes")
remotes::install_github("reconhub/aweek#21")
```

Below are some highlights:

You can now create aweek objects directly from data frames of week numbers\!

``` r
library("aweek")
library("dplyr")
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
set_week_start("Sunday") # Set the global week_start (defaults to Monday)
```

You can use numeric week data to create weeks or dates with `get_aweek()`
and `get_date()`. By default, they will give you the first week of the
current year:

``` r
get_aweek()
#> <aweek start: Sunday>
#> [1] "2019-W01-1"
get_date()
#> [1] "2018-12-30"
```

They are vectorised so you can use a data frame with them:

``` r
# create a table with all days in the week
d    <- as.Date("2019-03-03") + 0:6
res  <- lapply(weekdays(d), function(i) date2week(d, week_start = i))
resn <- lapply(weekdays(d), function(i) date2week(d, week_start = i, numeric = TRUE))
datf <- data.frame(wday = rep(weekdays(d), each = 7), 
                   week = unlist(res), # note: unlist converts to character
                   week_number = unlist(resn),
                   year  = 2019,
                   stringsAsFactors = FALSE)
datf$day <- substring(datf$week, 10, 11)
datf <- as_tibble(datf)
datf
#> # A tibble: 49 x 5
#>    wday   week       week_number  year day  
#>    <chr>  <chr>            <dbl> <dbl> <chr>
#>  1 Sunday 2019-W10-1          10  2019 1    
#>  2 Sunday 2019-W10-2          10  2019 2    
#>  3 Sunday 2019-W10-3          10  2019 3    
#>  4 Sunday 2019-W10-4          10  2019 4    
#>  5 Sunday 2019-W10-5          10  2019 5    
#>  6 Sunday 2019-W10-6          10  2019 6    
#>  7 Sunday 2019-W10-7          10  2019 7    
#>  8 Monday 2019-W09-7           9  2019 7    
#>  9 Monday 2019-W10-1          10  2019 1    
#> 10 Monday 2019-W10-2          10  2019 2    
#> # … with 39 more rows
```

Here you can convert week definitions with different starts to the correct
dates (the week of 2019-03-03).

``` r
datf %>%
  mutate(aweek = get_aweek(week = week_number, 
                           year = year, 
                           day = day, 
                           start = wday)) %>%
  mutate(the_date = get_date(week = week_number, 
                             year = year, 
                             day = day, 
                             start = wday))
#> # A tibble: 49 x 7
#>    wday   week       week_number  year day   aweek      the_date  
#>    <chr>  <chr>            <dbl> <dbl> <chr> <aweek>    <date>    
#>  1 Sunday 2019-W10-1          10  2019 1     2019-W10-1 2019-03-03
#>  2 Sunday 2019-W10-2          10  2019 2     2019-W10-2 2019-03-04
#>  3 Sunday 2019-W10-3          10  2019 3     2019-W10-3 2019-03-05
#>  4 Sunday 2019-W10-4          10  2019 4     2019-W10-4 2019-03-06
#>  5 Sunday 2019-W10-5          10  2019 5     2019-W10-5 2019-03-07
#>  6 Sunday 2019-W10-6          10  2019 6     2019-W10-6 2019-03-08
#>  7 Sunday 2019-W10-7          10  2019 7     2019-W10-7 2019-03-09
#>  8 Monday 2019-W09-7           9  2019 7     2019-W10-1 2019-03-03
#>  9 Monday 2019-W10-1          10  2019 1     2019-W10-2 2019-03-04
#> 10 Monday 2019-W10-2          10  2019 2     2019-W10-3 2019-03-05
#> # … with 39 more rows
```

Or, if you have ISO week formatted strings (YYYY-Www-d), then you can use
`as.aweek()`:

``` r
datf %>%
  mutate(aweek = as.aweek(week, start = wday)) %>%
  mutate(the_date = as.Date(aweek))
#> # A tibble: 49 x 7
#>    wday   week       week_number  year day   aweek      the_date  
#>    <chr>  <chr>            <dbl> <dbl> <chr> <aweek>    <date>    
#>  1 Sunday 2019-W10-1          10  2019 1     2019-W10-1 2019-03-03
#>  2 Sunday 2019-W10-2          10  2019 2     2019-W10-2 2019-03-04
#>  3 Sunday 2019-W10-3          10  2019 3     2019-W10-3 2019-03-05
#>  4 Sunday 2019-W10-4          10  2019 4     2019-W10-4 2019-03-06
#>  5 Sunday 2019-W10-5          10  2019 5     2019-W10-5 2019-03-07
#>  6 Sunday 2019-W10-6          10  2019 6     2019-W10-6 2019-03-08
#>  7 Sunday 2019-W10-7          10  2019 7     2019-W10-7 2019-03-09
#>  8 Monday 2019-W09-7           9  2019 7     2019-W10-1 2019-03-03
#>  9 Monday 2019-W10-1          10  2019 1     2019-W10-2 2019-03-04
#> 10 Monday 2019-W10-2          10  2019 2     2019-W10-3 2019-03-05
#> # … with 39 more rows
```

`change_week_start()` will allow you to re-calculate week numbers based on a
different week\_start attribute

``` r
w <- as.aweek(datf$week, start = datf$wday)
head(w)
#> <aweek start: Sunday>
#> [1] "2019-W10-1" "2019-W10-2" "2019-W10-3" "2019-W10-4" "2019-W10-5"
#> [6] "2019-W10-6"
head(change_week_start(w, "Monday"))
#> <aweek start: Monday>
#> [1] "2019-W09-7" "2019-W10-1" "2019-W10-2" "2019-W10-3" "2019-W10-4"
#> [6] "2019-W10-5"
```

`factor_aweek()` will aggregate your weeks with levels that span the range.

``` r
factor_aweek(as.aweek(as.Date(c("2019-03-03", "2019-05-31"))))
#> <aweek start: Sunday>
#> [1] 2019-W10 2019-W22
#> 13 Levels: 2019-W10 2019-W11 2019-W12 2019-W13 2019-W14 ... 2019-W22
```

<sup>Created on 2019-05-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

**Comment by zkamvar** [comment]
@scottyaz, since you've used aweek in the past, would you mind testing out this version?

```r
# install.packages("remotes")
remotes::install_github("reconhub/aweek#21")
```

**Comment by scottyaz** [comment]
Sure but it may be a few days.

**Comment by zkamvar** [comment]
> Sure but it may be a few days.

Thank You!

**Comment by dirkschumacher** [comment]
Just a quick note before I can look at it more in depth: I thought ISO weeks always start on Monday. Wouldn't a call to `set_week_start` invalidate all ISO weeks generated?

**Comment by zkamvar** [comment]
> Just a quick note before I can look at it more in depth: I thought ISO weeks always start on Monday. Wouldn't a call to `set_week_start` invalidate all ISO weeks generated?

Yes, ISO weeks all start on Monday, but all aweek objects contain their own `week_start` attribute, which only changes if the user specifies, so using `set_week_start()` will not invalidate any previously created `aweek` objects.

**Comment by dirkschumacher** [comment]
Installed and tested it. I don't have the time to go through the code really but it looks good to me and I really like the package.

**Review by dirkschumacher — APPROVED** [review]

File: DESCRIPTION [hunk 1]
```diff
@@ -11,6 +11,7 @@ Encoding: UTF-8
 LazyData: true
 Suggests: 
     testthat,
+    stats,
```
Comment (dirkschumacher) [inline DESCRIPTION:14]: Why?

Comment (zkamvar) [reply DESCRIPTION:14]: I use it in one of the tests: https://github.com/reconhub/aweek/pull/21/files#diff-ba0c75efb3eba3bc6b8b3e20055a0846R64

**Pull request merged**